app/setup.php

```diff
@@ -14,16 +14,22 @@
  * @return void
  */
 add_action('wp_enqueue_scripts', function () {
-    wp_enqueue_script('sage/vendor.js', asset('scripts/vendor.js')->uri(), ['jquery'], null, true);
-    wp_enqueue_script('sage/app.js', asset('scripts/app.js')->uri(), ['sage/vendor.js', 'jquery'], null, true);
+    $manifest = json_decode(asset('manifest.json')->contents(), true);
 
-    wp_add_inline_script('sage/vendor.js', asset('scripts/manifest.js')->contents(), 'before');
+    /**
+     * Needs working out.
+     */
+    wp_enqueue_script('sage/vendor/app', get_home_url() . $manifest['vendor/app.js'], ['jquery'], null, true);
+    wp_add_inline_script('sage/vendor/app', asset('runtime/app.js')->contents(), 'before');
+    wp_enqueue_script('sage/vendor/app~customizer', get_home_url() . $manifest['vendor/app~customizer.js'], ['sage/vendor/app'], null, true);
+    wp_enqueue_script('sage/vendor/app~editor~customizer', get_home_url() . $manifest['vendor/app~editor~customizer.js'], ['sage/vendor/app~customizer'], null, true);
+    wp_enqueue_script('sage/app', get_home_url() . $manifest['app.js'], ['sage/vendor/app~customizer'], null, true);
```

Comment:
It is not my intent for these changes to be permanent.

Ultimately, I think the entrypoints manifest should be readable from acorn and this space can be way more removed from the wordpress api (maybe something like `enqueueEntrypoint('app.js')`?) . Currently the entrypoints plugin generates this output:

**entrypoints.json**

```json
{
  "app": {
    "runtime/app": "/app/themes/sage/dist/runtime/app.72976a14.js",
    "vendor/app~customizer": "/app/themes/sage/dist/vendor/app~customizer.72976a14.js",
    "vendor/app": "/app/themes/sage/dist/vendor/app.72976a14.js",
    "app": "/app/themes/sage/dist/app.72976a14.css"
  },
  "editor": {
    "runtime/editor": "/app/themes/sage/dist/runtime/editor.72976a14.js",
    "editor": "/app/themes/sage/dist/editor.72976a14.css"
  },
  "customizer": {
    "runtime/customizer": "/app/themes/sage/dist/runtime/customizer.72976a14.js",
    "vendor/app~customizer": "/app/themes/sage/dist/vendor/app~customizer.72976a14.js",
    "customizer": "/app/themes/sage/dist/customizer.72976a14.js"
  }
}
```

Totally willing to change that if @QWp6t needs it structured differently.

Because of the proxy built into the dev server a missing asset won't return a 404 in development. This is why there are no acorn errors when running a dev build even though the css isn't written to disk.

Reply:
I think the best solution would be to put the filetypes in an array so we can loop it and remain using WordPress' enqueue functions.

From Encore:

```
{
  "app": {
    "js": [
      "modules/dashboard/runtime.js",
      "modules/dashboard/vendor.js",
      "modules/dashboard/app.js"
    ],
    "css": [
      "modules/dashboard/vendor.css",
      "modules/dashboard/app.css"
    ]
  }
}
```

Reply:
@QWp6t mentioned also wanting arrayed assets. how do people feel about this:

```json
{
  "entrypoints": {
    "app": {
      "js": {
        "runtime/app": "/app/themes/sage/dist/runtime/app.72976a14.js",
        "vendor/app~customizer": "/app/themes/sage/dist/vendor/app~customizer.72976a14.js",
        "vendor/app": "/app/themes/sage/dist/vendor/app.72976a14.js",
        "app": "/app/themes/sage/dist/app.72976a14.js"
      },
      "css": [
        "/app/themes/sage/dist/app.72976a14.css"
      ]
    }
  },
  "assets": [
    "/app/themes/sage/dist/images/image.png",
    "/app/themes/sage/dist/fonts/font.otf",
    "/app/themes/sage/dist/whatever-assets-arent-js-or-css.svg"
  ]
}
```

And, is it helpful to have absolute paths in addition to public root relative ones?

```json
{
  "entrypoints": {
    "app": {
      "js": {
        "runtime/app": {
          public: "/app/themes/sage/dist/runtime/app.72976a14.js",
          path: "/absolute/path/to/wp/app/themes/sage/dist/runtime/app.72976a14.js",
        }
      }
    }
  }
}
```

I could see that being most useful with the runtime chunks, since this would undo any need for the PHP to derive a path from the public URL.

Reply:
I'd say no to absolute path since it's a publicly readable file and I think it's weird exposing stuff like that about my web application – but maybe I'm weird. – otherwise the entrypoints looks 🔥

Reply:
For the assets, I was thinking of having them keyed by their relative path without cache-busting, similar to how we've always done it. And the CSS files should have a key so they can be named to ensure that if two entrypoints are loaded on a page and both share a CSS file, the CSS file will only be loaded once.

Also, I was hoping we could add the external dependencies for each entrypoint. I've already written something similar to work with encore so if you don't already have access to this, I can look into getting it into bud next week.

One more suggestion: the option to have the URL values be relative to the theme instead of root. This would allow developers to more easily distribute the theme separate from WordPress. That's what my reasoning behind it was in Sage 9.

In this example, I put the external dependencies in the entrypoints as well as a separate key on the bottom. We only need one, and it doesn't really matter which. So I'm cool with whatever. This is just to demonstrate what I'm talking about.

```json
{
  "entrypoints": {
    "app": {
      "js": {
        "runtime/app": "dist/runtime/app.72976a14.js",
        "vendor/app~customizer": "dist/vendor/app~customizer.72976a14.js",
        "vendor/app": "dist/vendor/app.72976a14.js",
        "app": "dist/app.72976a14.js"
      },
      "css": {
        "app": "dist/app.72976a14.css"
      },
      "dependencies": ["wp-blocks", "wp-dom-ready", "wp-element", "wp-hooks"]
    }
  },
  "assets": {
    "images/image.png": "dist/images/image.1a2b3c4d.png",
    "fonts/font.otf": "dist/fonts/font.1a2b3c4d.otf",
    "whatever-assets-arent-js-or-css.svg": "dist/whatever-assets-arent-js-or-css.1a2b3c4d.svg"
  },
  "dependencies": {
    "app": ["wp-blocks", "wp-dom-ready", "wp-element", "wp-hooks"]
  }
}
```

Let me know your thoughts. Like I mentioned in Slack, I'm totally fine if everyone isn't sold on the idea of having a unified manifest.json file.

Reply:
Done. Basically just implemented @Log1x's structure: https://github.com/roots/bud/pull/25

Output: 

```json
{
  "foo": {
    "js": [
      "/runtime/foo.js",
      "/vendor/lodash.js",
      "/foo.js"
    ],
    "css": ["/foo.css"],
    "dependencies": ["wp-element"]
  },
  "bar": {
    "js": [
      "/runtime/bar.js",
      "/vendor/lodash.js",
      "/bar.js"
    ],
    "css": [],
    "dependencies": ["wp-element"]
  }
}
```